30th-Assignment/30th-Assignment/30th-Assignment/Screens/Authentication/LoginViewController.swift

```diff
@@ -0,0 +1,156 @@
+    }
+
+    private lazy var signUpButton = UIButton().then {
+        $0.setTitle("가입하기", for: .normal)
+        $0.setTitleColor(.systemBlue, for: .normal)
+        $0.titleLabel?.font = .systemFont(ofSize: 17)
+
+        let pushSignUpViewAction = UIAction { _ in
+            self.navigationController?.pushViewController(MakeNameViewController(), animated: true)
+        }
+        $0.addAction(pushSignUpViewAction, for: .touchUpInside)
+    }
+
+    override func viewDidLoad() {
+        super.viewDidLoad()
+    }
+
+    override func viewWillAppear(_ animated: Bool) {
+        super.viewWillAppear(animated)
+        intialize()
+    }
+
+    /// 화면 터치했을 때 텍스트 필드 edit 종료하기
+    override func touchesBegan(_ touches: Set<UITouch>, with event: UIEvent?) {
+        view.endEditing(true)
+    }
+
+    override func configUI() {
+        setupBaseNavigationBar()
+        setTextField()
+    }
+
+    override func render() {
+        view.addSubViews([logoImage, emailTextField, passwordTextField, findPasswordButton, loginButton, signUpLabel, signUpButton])
+
+        logoImage.snp.makeConstraints {
+            $0.top.equalToSuperview().inset(100)
+            $0.centerX.equalToSuperview()
+        }
+
+        emailTextField.snp.makeConstraints {
+            $0.top.equalTo(logoImage.snp.bottom).offset(50)
+            $0.leading.trailing.equalToSuperview().inset(16)
+        }
+
+        passwordTextField.snp.makeConstraints {
+            $0.top.equalTo(emailTextField.snp.bottom).offset(10)
+            $0.leading.trailing.equalToSuperview().inset(16)
+        }
+
+        findPasswordButton.snp.makeConstraints {
+            $0.top.equalTo(passwordTextField.snp.bottom).offset(15)
+            $0.trailing.equalToSuperview().inset(16)
+        }
+
+        loginButton.snp.makeConstraints {
+            $0.top.equalTo(findPasswordButton.snp.bottom).offset(20)
+            $0.leading.trailing.equalToSuperview().inset(16)
+        }
+
+        signUpLabel.snp.makeConstraints {
+            $0.centerX.equalToSuperview()
+            $0.top.equalTo(loginButton.snp.bottom).offset(20)
+        }
+
+        signUpButton.snp.makeConstraints {
+            $0.centerY.equalTo(signUpLabel.snp.centerY)
+            $0.leading.equalTo(signUpLabel.snp.trailing).offset(10)
+        }
+    }
+
+    private func setTextField() {
+        [emailTextField, passwordTextField].forEach {
+            $0.delegate = self
+            $0.addTarget(self, action: #selector(textFieldDidChange(_:)), for: .editingChanged)
+        }
+    }
+
+    func intialize() {
+        /// 뷰 띄울때 텍필 초기화 하기 & 버튼 초기화
+        loginButton.isEnabled = false
+        [emailTextField, passwordTextField].forEach {
+            $0?.text = ""
+        }
+    }
+
+    @objc
+    private func textFieldDidChange(_ sender: InstaTextField) {
+        /// 도전과제 (2)
+        loginButton.isEnabled = (emailTextField.hasText && passwordTextField.hasText) ? true : false
```

Comment:
수꽐~~ and 연산자 반환값이 boolean 이니까 삼항 연산자 안써도 되지않나 싶네 큼큼 ㅎㅎ

```suggestion
        loginButton.isEnabled = emailTextField.hasText && passwordTextField.hasText
```

아니면 이런것도 해볼 수 있을듯!
```swift
loginButton.isEnabled = [emailTextField, passwordTextField].allSatisfy { $0.hasText }
```

Reply:
@noah0316  엥 쩐다 진짜 홍선배 천재 만재 개발자 ~~~~!!!!

Reply:
@noah0316  와 바로 적용해봐야지 ! allSatisfy 는 또 어디서 알아왔댜